### PR TITLE
A precision parameterised implementation of hyperloglog approximated cardinality for postgres and the django ORM

### PR DESCRIPTION
**Pull request by janrito**
Instead of counting the number of distinct users,

```python
Session.objects.aggregate(unique_users=Count("user_uuid", distinct=True))

{'unique_users': 71287}
```

you can approximate it,

```python
Session.objects.aggregate(approx_unique_users=ApproxCardinality("user_uuid"))

{'approx_unique_users': 76914}
```

Or, instead of counting the number of distinct users per day,

```python
list(
    Session.objects
        .annotate(date_of_session=TruncDate("created"))
        .values("date_of_session")
        .annotate(unique_users=Count("user_uuid", distinct=True))
        .values("unique_users", "date_of_session")
)

[{'date_of_session': datetime.date(2023, 4, 6), 'unique_users': 15292},
 {'date_of_session': datetime.date(2023, 4, 7), 'unique_users': 15289},
 {'date_of_session': datetime.date(2023, 4, 8), 'unique_users': 15395},
 {'date_of_session': datetime.date(2023, 4, 9), 'unique_users': 15484},
 {'date_of_session': datetime.date(2023, 4, 10), 'unique_users': 15315},
 {'date_of_session': datetime.date(2023, 4, 11), 'unique_users': 15346},
 {'date_of_session': datetime.date(2023, 4, 12), 'unique_users': 11791},
 {'date_of_session': datetime.date(2023, 4, 13), 'unique_users': 11672}]
```

you can approximate it,

```python

list(
    Session.objects
        .annotate(date_of_session=TruncDate("created"))
        .values("date_of_session")
        .annotate(approx_unique_users=ApproxCardinality("user_uuid"))
        .values("approx_unique_users", "date_of_session")
 )

 [{'date_of_session': datetime.date(2023, 4, 6), 'approx_unique_users': 15214},
 {'date_of_session': datetime.date(2023, 4, 7), 'approx_unique_users': 15686},
 {'date_of_session': datetime.date(2023, 4, 8), 'approx_unique_users': 15597},
 {'date_of_session': datetime.date(2023, 4, 9), 'approx_unique_users': 16056},
 {'date_of_session': datetime.date(2023, 4, 10), 'approx_unique_users': 15765},
 {'date_of_session': datetime.date(2023, 4, 11), 'approx_unique_users': 16620},
 {'date_of_session': datetime.date(2023, 4, 12), 'approx_unique_users': 12395},
 {'date_of_session': datetime.date(2023, 4, 13), 'approx_unique_users': 13029}]
```

The approximation is sometimes faster and uses less memory.

By default, it uses a precision of 9, which tends to have an error of about 5%. You can set a higher precision (up to ~11~ 31)) through a second parameter:

```python
list(
    Session.objects
        .annotate(date_of_session=TruncDate("created"))
        .values("date_of_session")
        .annotate(approx_unique_users=ApproxCardinality("user_uuid", 11))
        .values("approx_unique_users", "date_of_session")
)

[{'date_of_session': datetime.date(2023, 4, 6), 'approx_unique_users': 15430},
{'date_of_session': datetime.date(2023, 4, 7), 'approx_unique_users': 15261},
{'date_of_session': datetime.date(2023, 4, 8), 'approx_unique_users': 15739},
{'date_of_session': datetime.date(2023, 4, 9), 'approx_unique_users': 15357},
{'date_of_session': datetime.date(2023, 4, 10), 'approx_unique_users': 15511},
{'date_of_session': datetime.date(2023, 4, 11), 'approx_unique_users': 15117},
{'date_of_session': datetime.date(2023, 4, 12), 'approx_unique_users': 11658},
{'date_of_session': datetime.date(2023, 4, 13), 'approx_unique_users': 11841}]
```

You can also set a lower precision (down to ~3~ 4) for a faster, less accurate estimate.

The aggregation is also available in SQL:

```sql

select
    date_trunc('day', created) as date_of_session,
    hll_approx_cardinality(user_uuid, 11) as approx_unique_users
from
    testapp_session
group by
    date_trunc('day', created);
    date_of_session     | approx_unique_users
------------------------+---------------------
 2023-04-06 00:00:00+00 |               15430
 2023-04-07 00:00:00+00 |               15261
 2023-04-08 00:00:00+00 |               15739
 2023-04-09 00:00:00+00 |               15357
 2023-04-10 00:00:00+00 |               15511
 2023-04-11 00:00:00+00 |               15117
 2023-04-12 00:00:00+00 |               11658
 2023-04-13 00:00:00+00 |               11841
(8 rows)

Time: 1289.249 ms (00:01.289)
```

### CODE REVIEW

**Comment by elliott-omosheye** [comment]
Looks great. Could also be good to show performance over size of the input set.

**Comment by janrito** [comment]
> Looks great. Could also be good to show performance over size of the input set.

yeah, good idea 45a964a999dbe54c978f4d49d21bc10cbcdc9d50

File: django_pg_simple_hll/migrations/0001_initial.sql [hunk 1]
```diff
@@ -0,0 +1,145 @@
+-- The state transition function
+-- it takes the current state, and the two input values
+-- hll_agg_state is the current running state, made up of bucket keys and hashes
+--  it is an array of n_buckets + 2
+-- input is the element we are considering
```
Comment (sam-lawrence) [inline django_pg_simple_hll/migrations/0001_initial.sql:5]: element as in a value from the column we're considering?

File: django_pg_simple_hll/migrations/0001_initial.sql [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+--  it use used to calculate the number of buckets
+create or replace function
+hll_bucket(hll_agg_state int[], input anyelement, hll_precision int)
+returns int[]
+language plpgsql immutable strict parallel safe as
+$$
+declare
+    n_buckets int := POW(2, hll_precision); -- eg.. for 512 buckets, hll_precision is 9 because 2**9 -> 512
+    hashed_input int := hashtext(input::text); -- transform the input into a text and hash it
+    bucket_hash int := hashed_input & ~(1 << 31); -- keep all the significant bits, without the sign
```
Comment (sam-lawrence) [inline django_pg_simple_hll/migrations/0001_initial.sql:16]: I'm possibly reading this wrong.  But does this mean if `hashed_input` looks like `(b1)(b2) (b3)...(bn)` were each b1..bn is a bit, then `bucket_hash` will look like `0(b2)(b3)...(b31)00...00`?

Comment (sam-lawrence) [reply django_pg_simple_hll/migrations/0001_initial.sql:16]: or actually perhaps I mean `bucket_hash` will look like `(b n-30)(b n-29)...b(n)`.  You just keep the last 31 bits?

File: django_pg_simple_hll/migrations/0001_initial.sql [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    bucket_key int := hashed_input & (n_buckets - 1); -- bucket the hash into one of n buckets
```
Comment (sam-lawrence) [inline django_pg_simple_hll/migrations/0001_initial.sql:17]: So only the last `hll_precision` bits of `bucket_key` actually matter?

File: django_pg_simple_hll/migrations/0001_initial.sql [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    -- pre fetch the hash for occupying the corresponding bucket for the input
+    -- we add 2 because
+    --  1. postgres arrays are 1-indexed
+    --  2. we have to account for an extra first element
+    current_hash int := hll_agg_state[bucket_key + 2];
+begin
+    -- we can only handle precision up to 11 or 2048 buckets
+    -- because hashtext produces a hash of int size with 31 useable bits
```
Comment (sam-lawrence) [inline django_pg_simple_hll/migrations/0001_initial.sql:25]: This is again my naive sense of what's going on, but i would have thought the precision could go up to 31 if "hashtext produces a hash of int size with 31 useable bits"

File: django_pg_simple_hll/migrations/0001_initial.sql [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    if hll_precision < 3 or hll_precision > 11 then
+        raise exception 'invalid hll_precision: % - must be between 3 and 11 inclusive', hll_precision;
+    end if;
+    -- postgres squeezes null elements in the array, we don't want that,
```
Comment (sam-lawrence) [inline django_pg_simple_hll/migrations/0001_initial.sql:29]: As in it removes null values?

File: django_pg_simple_hll/migrations/0001_initial.sql [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+-- The finalfunc
+-- takes the hll_agg_state and approximates cardinality
+create or replace function
+hll_approximate(hll_agg_state int[])
+returns int
+language sql immutable as
+$$
+with hll_agg_state_table as (
+    select
+        31 - FLOOR(LOG(2, bucket_hash)) as bucket_hash,
```
Comment (sam-lawrence) [inline django_pg_simple_hll/migrations/0001_initial.sql:82]:
```
as bucket_hash
```
This seems quite different from the `bucket_hash` now, and from a hash in general.  Why rename rather than giving a new name?  What's the significance of this number?

File: django_pg_simple_hll/migrations/0001_initial.sql [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        (array_length(hll_agg_state, 1) - 2) as n_buckets
+    from unnest(hll_agg_state) with ordinality as hll_agg_state_table(bucket_hash, bucket_key)
+    where bucket_hash is not null -- ignore all null elements
+        and bucket_key != 1 -- ignore the first element of the array
+        and bucket_key != array_length(hll_agg_state, 1) -- ignore the last element of the array
+),
+-- compute counts and aggregates
+counted as (
+    select
+        MAX(hll_agg_state_table.n_buckets) as n_buckets,
```
Comment (sam-lawrence) [inline django_pg_simple_hll/migrations/0001_initial.sql:93]: Seems like it'd be simpler to pass around the `n_buckets` argument than storing it in `hll_agg_state`?

File: django_pg_simple_hll/migrations/0001_initial.sql [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        MAX(hll_agg_state_table.n_buckets) - COUNT(hll_agg_state_table.bucket_hash) as n_zero_buckets,
+        SUM(POW(2, -1 * hll_agg_state_table.bucket_hash)) as harmonic_mean
+    from hll_agg_state_table
+),
+-- estimate
+estimation as (
+    select
+        (
+            (POW(counted.n_buckets, 2) * (0.7213 / (1 + 1.079 / counted.n_buckets))) /
```
Comment (sam-lawrence) [inline django_pg_simple_hll/migrations/0001_initial.sql:102]: `0.7213`, `1.079`.  Would be great to have a comment for these numbers

**Review by sam-lawrence — APPROVED** [review]
Looks really cool.  Can't say that I was fully able to follow the logic, but I'm entirely unfamiliar with the algorithm

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 2 above)
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql:16]: this keeps the rightmost 31 bits, and sets the "sign" bit to 1
`1(b1)(b2)...(b32)`

In python 
```python
f"{-1 & ~(1 << 31) & 0xffffffff:0b}" -> '1111111111111111111111111111111'
f"{(2**32 -1)  & ~(1 << 31) & 0xffffffff:0b}" -> '1111111111111111111111111111111'
```

(python int's have variable length that's why you have to do the `& 0xffffffff` bit

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 3 above)
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql:17]: yes:
```python
f"{(9)  & (2**3 - 1) & 0xffffffff:0b}" -> 1
````

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 5 above)
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql:29]: it does somethign weird: if an array has null values at the beginning or the end, it will just reduce the size of the array to fit. This woud be fine if it would be at the end, but it also does it at the beginning

```
{1, Null, 3, Null, Null} -> {1, Null, 3}
{Null, Null, 1, Null, 3, Null, Null} -> {1, Null, 3}
```

So, what we do is we pre-alllocate an array of N+2 values and set the first and last element to a constant:
```
{4, Null, Null, Null, Null, 4}
```

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 4 above)
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql]: you are right. I think I confused an earlier attempt to make the hash 64 bits long 

I've corrected it here e305c7289c992f6d59c9c382b7ed9e20770b118d

**Comment by janrito** [comment]
I'm giving up trying to find a statistical test for this.

I still would like to have a way to unit test it, though.  

The one idea I had is to test that the average error of an estimate at precision `x` is less than the average error of the estimate at precision `x - 1`?

What do you think?

**Comment by sam-lawrence** [comment]
> I'm giving up trying to find a statistical test for this.
> 
> I still would like to have a way to unit test it, though.
> 
> The one idea I had is to test that the average error of an estimate at precision `x` is less than the average error of the estimate at precision `x - 1`?
> 
> What do you think?

I'm not against this but I would have thought you run into exactly the same problem in that it's still a statistical test?  What's the difference between a test like this and a test that the estimate is within a certain range?

**Comment by janrito** [comment]
> > I'm giving up trying to find a statistical test for this.
> > I still would like to have a way to unit test it, though.
> > The one idea I had is to test that the average error of an estimate at precision `x` is less than the average error of the estimate at precision `x - 1`?
> > What do you think?
> 
> I'm not against this but I would have thought you run into exactly the same problem in that it's still a statistical test? What's the difference between a test like this and a test that the estimate is within a certain range?

how would ypu pick a range?

File: django_pg_simple_hll/migrations/0001_initial.sql [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    hll_agg_state int[],
+    input anyelement,
+    hll_precision int
+) returns int[]
+language plpgsql immutable strict parallel safe as $$
+declare
+    -- eg. for 512 buckets, hll_precision is 9 because 2**9 -> 512
+    n_buckets int := POW(2, hll_precision);
+    -- transform the input into a text and hash it
+    hashed_input int := hashtext(input::text);
```
Comment (alexhenman) [inline django_pg_simple_hll/migrations/0001_initial.sql:18]: So `hashtext` is an undocumented internal postgres function: are there any particular properties we need of the hash function for this aggregate to work correctly?

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 5 above)
Comment (alexhenman) [reply django_pg_simple_hll/migrations/0001_initial.sql:35]: Is there a specific function you're calling that's doing this? I can't see any reference to this behaviour and this doesn't appear to be true just creating an array in a postgres shell:

```SQL
master@host:ukf> select '{NULL, 1, NULL, 3, NULL, NULL}';
+--------------------------------+
| ?column?                       |
|--------------------------------|
| {NULL, 1, NULL, 3, NULL, NULL} |
+--------------------------------+
SELECT 1
Time: 0.014s
```

File: django_pg_simple_hll/migrations/0001_initial.sql [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    end if;
+
+    if current_hash is null or current_hash > bucket_hash then
+        hll_agg_state[bucket_key + 2] := bucket_hash;
+    end if;
+    return hll_agg_state;
+end $$;
+
+
+-- sfunc with default precision of 9
```
Comment (alexhenman) [inline django_pg_simple_hll/migrations/0001_initial.sql:49]: Can you do this by just providing a default for the parameter in the above function definition?

File: testapp/tests.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+import pytest
+from django.db.models import Count
+from django.db.models.functions import TruncDate
+from django_pg_simple_hll.aggregate import ApproxCardinality
+
+from .models import Session
+
+FIELDS = ("user_int", "user_uuid", "user_str")
+PRECISIONS_TO_TEST = (5, 8, 9, 10, 11, 12)  # if tails intermittently with 4
```
Comment (alexhenman) [inline testapp/tests.py:13]:
```suggestion
PRECISIONS_TO_TEST = (5, 8, 9, 10, 11, 12)  # it fails intermittently with 4
```

Does this mean we're not testing the `m=16` and `m=64` cases of the alpha_m approximation constant?

If we're getting intermittent failures do we just need to re-run the test more times to get a better average of averages?

**Comment by sam-lawrence** [comment]
> > > I'm giving up trying to find a statistical test for this.
> > > I still would like to have a way to unit test it, though.
> > > The one idea I had is to test that the average error of an estimate at precision `x` is less than the average error of the estimate at precision `x - 1`?
> > > What do you think?
> > 
> > 
> > I'm not against this but I would have thought you run into exactly the same problem in that it's still a statistical test? What's the difference between a test like this and a test that the estimate is within a certain range?
> 
> how would you pick a range?

Somewhat arbitrarily, but I don't think that's particularly different to what's happening here.  I guess pick a range such there's a very low probability hyperloglog should fall outside of it, but narrow enough to ensure that it seems to be estimating the right thing.

That's not ideal, and I don't necessarily think it's better than comparing estimates with different parameters, but I also don't think it's really much worse.  Seems to me the issues are the same?

We're not actually interested in testing how accurate this algorithm is though right?  Just that this is a correct implementation of the algorithm?   Can we set up some initial conditions such that we know we should give a specific estimate value from it?  That might be a more reliable test.  You could even compare it to a python implementation of hyperloglog on the same data

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 9 above)
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql:18]: Yes. It needs to produce uniformly distributed integers from the input, which is exactly what `hashtext` produces.

The majority of the cost of the aggregation is hashing, so it is important that it is fast. 

We could use something like md5 or `sha256`, which have some interesting properties:
- we could generate 64 bit hashes, which will improve the range of cardinalities that we can estimate
- well documented & reproducible in python

However, they produce strings, which need to be cast to integers. This makes the operation considerably slower.


`Hashtext` is undocumented - there have been some questions about whether it should be documented: https://www.postgresql.org/message-id/24463.1329854466%40sss.pgh.pa.us with limited support. Partially because they don't want to guarantee that the current behaviour will be maintained. In that same message, it is suggested that it could be changed to produce 64 bit hashes :)

For our purposes, though, we only care that the results are fast and consistent within the current session.

File: django_pg_simple_hll/migrations/0001_initial.reverse.sql [hunk 12]
```diff
@@ -0,0 +1,6 @@
+drop aggregate if exists hll_approx_cardinality(anyelement, int);
```
[the remark this answers is not in the document]
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.reverse.sql:1]: https://docs.sqlfluff.com/en/stable/rules.html#capitalisation-bundle

[the remark this answers is not in the document]
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.reverse.sql:1]: I'm not enforcing sqlfluff, because there is an open issue with `create aggregate` https://github.com/sqlfluff/sqlfluff/issues/3556

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 9 above)
Comment (alexhenman) [reply django_pg_simple_hll/migrations/0001_initial.sql:18]: OK just wanted to check that "produce uniformly distributed integers" bit is something we've checked rather than just an assumption. Since it's an internal function it feels like it could be designed to produce unique hashes but not necessarily such that a small change in input results in a big change in the hash

**Comment by elliott-omosheye** [comment]
Yea thats my same though. The results are deterministic so we should just be ensuring that the sql version gives the correct answer for a given dataset vs a known value or another implementation.

File: django_pg_simple_hll/migrations/0001_initial.reverse.sql (hunk 12 above)
[the remark this answers is not in the document]
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.reverse.sql:1]: have tried to format it consistently here 44598351cf6f3c5194f6cfa96f7be48d19f0f9de and add some non-enforced linting here 8d775c1525e4c20fd80ea7d039c062b57853b042

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 10 above)
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql:49]: I've tried to do this, but can't manage it. Not sure if there is a trick

File: django_pg_simple_hll/aggregate.py [hunk 13]
```diff
@@ -0,0 +1,7 @@
+from django.db.models import Aggregate, IntegerField
+
+
+class ApproxCardinality(Aggregate):
```
[the remark this answers is not in the document]
Comment (janrito) [reply django_pg_simple_hll/aggregate.py:4]: I've added some of the other properties like Count. 

The init function is probably not relevant. The aggregation wont work on `*` input, only on a specific column.

In terms of language, I'm not sure that I would want to change it to Count. The algorithm is specific for approximating cardinality – the number of distinct elements in a set – not counting them.

[the remark this answers is not in the document]
Comment (alexhenman) [reply django_pg_simple_hll/aggregate.py:4]: Yea I'd be inclined to stick to postgres language though really since the name is more for the technical user who shouldn't really need to be worrying about the exact hyperloglog implementation or set theory.

There also seems to be some ambiguity about the definition of cardinality of a multiset. Wiki seems to think that actually doesn't mean the distinct count at all:

> The cardinality of a multiset is the sum of the multiplicities of all its elements. For example, in the multiset {a, a, b, b, b, c} the multiplicities of the members a, b, and c are respectively 2, 3, and 1, and therefore the cardinality of this multiset is 6.

Whereas other sources seem to think the cardinality there would be 3.

I think `DistinctCount` is probably a bit clearer and more in plain English

[the remark this answers is not in the document]
Comment (sam-lawrence) [reply django_pg_simple_hll/aggregate.py:4]: > The algorithm is specific for approximating cardinality – the number of distinct elements in a set – not counting them.

Somewhat along the lines of what Alex has said, I'm not sure this is quite what cardinality is, at least in the classic mathematical setting.  

The motivation for cardinality was not to have an alternative term for distinct counts, but to able extend the concept of "size" in set theory from a count (which would be distinct anyway) of finite sets to a measure of infinite sets.  I don't think "distinctness" is really vital to the idea of cardinality -  it's just implicit because cardinality is associated with sets.  Since we're not dealing with sets here anyway I think it probably is worth being explicit about the distinctness.

I don't actually have a strong opinion on the name, I think it's fairly obvious what you mean. I just wanted to get invovled in the chat. Cantor's proofs got me into mathematics.

[the remark this answers is not in the document]
Comment (janrito) [reply django_pg_simple_hll/aggregate.py:4]: I should have known not to explain terminology with mathematicians.

However, I don't think that I am making a wild claim here. It looks like cardinality is the common language used for this exact operation in databases: 
postgres: https://github.com/citusdata/postgresql-hll 
redis: https://redis.io/docs/data-types/hyperloglogs/ 
elasticsearch: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-cardinality-aggregation.html 

And it is the way that it is called in the original research papers:
http://algo.inria.fr/flajolet/Publications/FlMa85.pdf
http://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 5 above)
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql:35]: You are right, it has to do wih a function:
```sql
SELECT * FROM UNNEST('{NULL, NULL, 1, NULL}'::int []);
--------
 [NULL]
 [NULL]
      1
 [NULL]
(4 rows)

Time: 1.561 ms
```

with ordinality:

```sql
SELECT v, idx FROM UNNEST('{NULL, NULL, 1, NULL}'::int []) WITH ORDINALITY AS arr(v, idx);
   v    | idx
--------+-----
 [NULL] |   1
 [NULL] |   2
      1 |   3
 [NULL] |   4
(4 rows)

Time: 3.958 ms
```

and even aggregating:

```sql
SELECT ARRAY_AGG(v) FROM UNNEST('{NULL, NULL, 1, NULL}'::int []) WITH ORDINALITY AS arr(v, idx);
     array_agg
--------------------
 {NULL,NULL,1,NULL}
(1 row)

Time: 10.252 ms
```

but as soon as it is the output of a function

```sql
SELECT * FROM UNNEST(hll_bucket('{}'::int [], 'hello'::text, 6));
  unnest
-----------
 277190697
(1 row)

Time: 1.074 ms
```

or in more detail

```sql
SELECT * FROM UNNEST(hll_bucket('{}'::int [], 'hello'::text, 6)) WITH ORDINALITY AS arr(v, idx);
     v     | idx
-----------+-----
 277190697 |   1
(1 row)

Time: 9.225 ms
```

Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql:37]: And with the _flanking_:
```sql
SELECT * FROM hll_bucket('{}'::int[], 'hello'::text, 4);
                                          hll_bucket
----------------------------------------------------------------------------------------------
 {16,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,277190697,NULL,NULL,NULL,NULL,NULL,NULL,16}
(1 row)

Time: 3.907 ms
```

unnesting

```sql
SELECT * FROM UNNEST(hll_bucket('{}'::int[], 'hello'::text, 4)) WITH ORDINALITY AS arr(v, idx);
     v     | idx
-----------+-----
        16 |   1
    [NULL] |   2
    [NULL] |   3
    [NULL] |   4
    [NULL] |   5
    [NULL] |   6
    [NULL] |   7
    [NULL] |   8
    [NULL] |   9
    [NULL] |  10
 277190697 |  11
    [NULL] |  12
    [NULL] |  13
    [NULL] |  14
    [NULL] |  15
    [NULL] |  16
    [NULL] |  17
        16 |  18
(18 rows)

Time: 15.465 ms
```

File: django_pg_simple_hll/aggregate.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (alexhenman) [reply django_pg_simple_hll/aggregate.py:4]: I think cardinality is fine if you like long words over count but I'd still say the distinct bit needs to be explicit unless you're going to explicitly reference in the name that this uses the HLL algorithm.

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 5 above)
Comment (alexhenman) [reply django_pg_simple_hll/migrations/0001_initial.sql:37]: Slightly struggling to follow where you're saying you think the issue is. Might be clearer to see if you demo with some dummy functions. I don't think it's as simple as just "it has to do with it being the output of a function":

```SQL
master@host:ukf> create or replace function dummy_return_array() returns int [] language plpgsql immutable strict parallel safe as $$ begin return '{NULL, 1, NULL}'; end $$;
CREATE FUNCTION
Time: 0.002s
master@host:ukf> select dummy_return_array();
+--------------------+
| dummy_return_array |
|--------------------|
| [None, 1, None]    |
+--------------------+
SELECT 1
Time: 0.005s
master@host:ukf> select * from unnest(dummy_return_array()) with ordinality as arr(v, idx);
+--------+-----+
| v      | idx |
|--------+-----|
| <null> | 1   |
| 1      | 2   |
| <null> | 3   |
+--------+-----+
```

Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql:37]: you are right, maybe it's about the direct assignment?
```sql
CREATE OR REPLACE FUNCTION dummy_return_array(s int [])
RETURNS int []
LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE AS
$$
begin
    s[3] := 1;
    s[5] := 2;
    return s;
end $$;

SELECT
    v,
    idx
FROM unnest(dummy_return_array('{}')) WITH ORDINALITY AS arr (v, idx);
CREATE FUNCTION
Time: 5.429 ms
   v    | idx
--------+-----
      1 |   1
 [NULL] |   2
      2 |   3
(3 rows)

Time: 1.637 ms```

Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql:37]: we could probably achieve the same thing assigning the end and the beginning of the array
```sql
CREATE OR REPLACE FUNCTION dummy_return_array(s int [], p int DEFAULT 3)
RETURNS int []
LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE AS
$$
begin
    s[1] := coalesce(s[1], NULL);
    s[3] := 1;
    s[5] := 2;
    s[pow(2, p)] := coalesce(s[1], NULL);
    return s;
end $$;

SELECT
    v,
    idx
FROM unnest(dummy_return_array('{}')) WITH ORDINALITY AS arr (v, idx);
CREATE FUNCTION
Time: 11.999 ms
   v    | idx
--------+-----
 [NULL] |   1
 [NULL] |   2
      1 |   3
 [NULL] |   4
      2 |   5
 [NULL] |   6
 [NULL] |   7
 [NULL] |   8
(8 rows)

Time: 4.463 ms```

File: django_pg_simple_hll/aggregate.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (kevinmarsh) [reply django_pg_simple_hll/aggregate.py:4]: My 2cents on the name is that something like `ApproximateCount` is easier for devs not familiar with HLL to understand, however if you don't understand the drawbacks of of this aggregation then you probably shouldn't be using it? Also this aggregate is very tightly coupled to the HLL implementation (eg. we wouldn't swap this out with [another approximate count algorithm](https://web.archive.org/web/20230321011643/https://wiki.postgresql.org/wiki/Count_estimate)) so I think it's fine if it uses language more associated with HLL.

Maybe a docstring could be a good compromise to help with IDEs if someone comes across this for the first time in a codebase rather than viewing this source repo?
```suggestion
class ApproxCardinality(Aggregate):
    """
    Return an approximate count based on the HyperLogLog algorithm.
    
    This should be much faster than Django's Count but will give
    an approximate result rather than an exact count.
    """
```

**Review by kevinmarsh — COMMENTED** [review]
I've done a quick review more from the POV of end user rather than on correctness of the algorithm. Is looking good though :+1:

**Comment by janrito** [comment]
> Yea thats my same thought. The results are deterministic so we should just be ensuring that the sql version gives the correct answer for a given dataset vs a known value or another implementation.

I've been thinking about this. I think the issue is that we need to have the same implementation of the algorithm, including the hashing. And the hashing is postgres specific

File: pyproject.toml [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+]
+packages = [
+  { include = "django_pg_simple_hll" },
+  { include = "django_pg_simple_hll/py.typed" },
+]
+
+[tool.poetry.dependencies]
+python = ">=3.9,<3.12"
+Django = ">=3.2.4"
+psycopg2 = "^2.9.6"
```
[the remark this answers is not in the document]
Comment (janrito) [reply pyproject.toml:35]: have put this in an issue #2

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 5 above)
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql]: da160e6c4b80b25596159575680ba690111cc5bf this works better

File: django_pg_simple_hll/migrations/0001_initial.sql (hunk 10 above)
Comment (janrito) [reply django_pg_simple_hll/migrations/0001_initial.sql:49]: I don't think this is possible. 

`CREATE AGGREGATE's` assumes it has the same singature than that of the `SFUNC`. So, we could setup a default argument to `hll_bucket`

```sql
CREATE OR REPLACE FUNCTION hll_bucket(
    hll_agg_state int [],
    input anyelement,
    hll_precision int DEFAULT 9
) RETURNS int []
...
```

but we cannot provide a default for the `AGGREGATE`.

File: testapp/tests.py (hunk 11 above)
Comment (janrito) [reply testapp/tests.py]: added back here: 060374d5cb36abde59482ff8929302dc6442a3aa

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+jobs:
+  build-and-test:
+    runs-on: ubuntu-latest
+    name: "CI | python v${{ matrix.python-version }}, postgres v${{ matrix.postgres-version }}"
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.9", "3.10", "3.11"]
+        postgres-version: ["12", "13", "14"]
```
[the remark this answers is not in the document]
Comment (janrito) [reply .github/workflows/ci.yml]: great idea 5fceb4a029cf77976aaff9f2d0ed67222b477be6

File: testapp/tests.py
[the remark this answers is not in the document]
Comment (janrito) [reply testapp/tests.py:1]: Great points, i've incorporated these here 59ab35771e4abafca36ac7b1fd26873f2c0fd2ab 


Any other tests, we could add as a separate PR

File: django_pg_simple_hll/aggregate.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (janrito) [reply django_pg_simple_hll/aggregate.py]: > > The aggregation wont work on `*` input, only on a specific column
> 
> By "won't work" do you mean Django/Posgres will raise an exception or it will give inaccurate/nonsensical results? If it's the latter then we should catch that to prevent bad data being returned

It will raise errors. I've added tests here 59ab35771e4abafca36ac7b1fd26873f2c0fd2ab

File: pyproject.toml [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+license = "MIT"
+readme = "README.md"
+classifiers = [
+  "Environment :: Web Environment",
+  "Framework :: Django",
+  "Framework :: Django :: 3.0",
+  "Framework :: Django :: 3.1",
+  "Framework :: Django :: 3.2",
+  "Framework :: Django :: 4.0",
+  "Framework :: Django :: 4.1",
```
[the remark this answers is not in the document]
Comment (janrito) [reply pyproject.toml]: yep: bce5f411ef5eb47b94ca3ad18813559bbbc66c86 and tests cc85cae4e36072f4cfe59aa60468ec591789bed1

File: .github/workflows/ci.yml [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+jobs:
+  build-and-test:
+    runs-on: ubuntu-latest
+    name: "CI | python v${{ matrix.python-version }}, postgres v${{ matrix.postgres-version }}"
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.9", "3.10", "3.11"]
+        postgres-version: ["12", "13", "14", "15"]
+        django-version: ["3.2", "4.0", "4.1", "4.2"]
```
Comment (kevinmarsh) [inline .github/workflows/ci.yml:17]: Sorry just one more bit of feedback on the testing matrix, you'll probably want to add a couple "excludes" here since for instance [Django 3.2 doesn't run on Python 3.11](https://docs.djangoproject.com/en/4.2/faq/install/#what-python-version-can-i-use-with-django) 



```suggestion
        django-version: ["3.2", "4.0", "4.1", "4.2"]
        exclude:
          - python-version: "3.11"
            django-version: "3.2"
          - python-version: "3.11"
            django-version: "4.0"
```

You can see an [example of excludes here](https://github.com/kevinmarsh/django-classy-tags/blob/021706f911920e84b43d646b0a50837ea3add156/.github/workflows/test.yml#L11-L21)

Comment (janrito) [reply .github/workflows/ci.yml:17]: 👍🏽

**Comment by janrito** [comment]
Thanks @elliott-omosheye @kevinmarsh @sam-lawrence @alexhenman for advice and comments
